File: github_code/lru-cache/java/1.java
package com.github.spbennett;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * Created by sbennett on 7/9/15.
 *
 * LRU Cache implementation in Java
 * Cache maps IDs to Strings.
 */
public class LRU {
    final int CAPACITY = 5;
    ConcurrentHashMap<Integer, String> cache = new ConcurrentHashMap<>(CAPACITY);
    Queue<Integer> lru = new PriorityQueue<>(CAPACITY);

    /**
     * Looks up an item in the cache
     * @param id Integer ID of the item to find in cache.
     * @return Object from cache or null if
     * not in cache.
     */
    String retrieve(Integer id){
        if (!cache.containsKey(id)){
            // Item not in cache.
            return null;
        } else {
            // Item found in cache.
            // Update the LRU and return the item.
            lru.remove(id);
            lru.add(id);
            return cache.get(id);
        }
    }

    void insert(Integer id, String value){
        if (cache.size() == CAPACITY){
            // Lookup LRU and remove it from cache to make
            // room.
            Integer markedForDelete = lru.remove();
            cache.remove(markedForDelete);
        }

        // Add item to cache and update LRU
        cache.put(id, value);
        lru.add(id);
    }

    public static void main (String args[]){
        LRU lru = new LRU();

        for (int i = 0; i < lru.CAPACITY; i++){
            // populate cache
            lru.insert(i, Integer.toString(i));
        }
        //lru.insert(new Integer(1), "ONE");

        System.out.println(lru.retrieve(1));

    }
}

File: github_code/lru-cache/cpp/1.c
/*
    Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and set.
		get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
		set(key, value) - Set or insert the value if the key is not already present. When the cache reached its capacity, 
		it should invalidate the least recently used item before inserting a new item.

    Link: http://www.lintcode.com/en/problem/lru-cache/

    Example: None

    Solution: Use a list and a map to implement get and set key,value information. 

    Source: https://github.com/kamyu104/LintCode/blob/master/C%2B%2B/lru-cache.cpp
*/

#include <list>

class LRUCache{
public:
    // @param capacity, an integer
    LRUCache(int capacity) : capa_(capacity) {
        // write your code here
    }
    
    // @return an integer
    int get(int key) {
        // write your code here
        auto it = map_.find(key);
        if (it != map_.end()) {
            // It key exists, update it.
            auto l_it = it->second;
            int value = l_it->second;
            update(it, value);
            return value;
        } else {
            return -1;
        }
    }

    // @param key, an integer
    // @param value, an integer
    // @return nothing
    void set(int key, int value) {
        // write your code here
        auto it = map_.find(key);
        // It key exists, update it.
        if (it != map_.end()) {
            update(it, value);
        } else {
            // If cache is full, remove the last one.
            if (list_.size() == capa_) {
                auto del = list_.back(); list_.pop_back();
                map_.erase(del.first);
            }
            list_.emplace_front(key, value);
            map_[key]=list_.begin();
        }
    }
    
private:
    list<pair<int, int>> list_; // key, value
    unordered_map<int, list< pair<int, int>>::iterator> map_; // key, list iterator
    int capa_;

    // Update (key, iterator of (key, value)) pair
    void update(unordered_map<int, list<pair<int, int>>::iterator>::iterator it, int value) {
        auto l_it = it->second;
        int key = l_it->first;
        list_.erase(l_it);
        list_.emplace_front(key, value);
        it->second = list_.begin();
    }
};